### PR TITLE
Refactor IoContext::Scope and ThreadScope into callback-based API

### PR DESCRIPTION
**Pull request by jasnell**
Continued work refactoring the Lock-taking APIs to a callback-style API .... There's still more to do but this PR is large enough already. Will keep working in increments

### CODE REVIEW

File: src/workerd/jsg/setup.h [hunk 1]
```diff
@@ -431,6 +434,15 @@ class Isolate: public IsolateBase {
     Isolate& jsgIsolate;
   };
 
+  void runInLockScope(auto func) {
```
Comment (mikea) [inline src/workerd/jsg/setup.h:437]: please, please add some comment about `func` signature! I've spent so much time deducing such things in our codebase...

File: src/workerd/io/worker.h [hunk 2]
```diff
@@ -549,6 +552,13 @@ class Worker::LockType {
   friend class Worker::Isolate;
 };
 
+auto Worker::runInLockScope(LockType lockType, auto func) const {
```
Comment (mikea) [inline src/workerd/io/worker.h:555]: please document func signature.

**Review by mikea — APPROVED** [review]
lgtm, removes lots of boilerplate

**Comment by jasnell** [comment]
> ... removes lots of boilerplate

And more to come :-) .

File: src/workerd/api/crypto-impl-aes-test.c++ [hunk 3]
```diff
@@ -96,16 +94,14 @@ KJ_TEST("AES-CTR key wrap") {
...
-    CryptoIsolate::Lock isolateLock(cryptoIsolate, stackScope);
+  e.getIsolate().runInLockScope([&](CryptoIsolate::Lock& isolateLock) {
     auto isolate = isolateLock.v8Isolate;
 
     isolateLock.withinHandleScope([&] {
       auto context = isolateLock.newContext<CryptoContext>().getHandle(isolateLock);
       auto contextScope = isolateLock.enterContextScope(context);
 
-      auto& js = jsg::Lock::from(isolate);
+      jsg::Lock& js = isolateLock;
```
Comment (fhanau) [inline src/workerd/api/crypto-impl-aes-test.c++:104]: Nit: With this being done, can the remaining instances of `jsg::Lock::from(isolate)` also be cleaned up or is `Lock::from` needed there?

Comment (jasnell) [reply src/workerd/api/crypto-impl-aes-test.c++:104]: Much of this is getting cleaned up in follow up PRs. Stay tuned.

**Pull request merged**